app/services/candidate_courses_recommender.rb

```diff
@@ -142,4 +143,26 @@ def subjects
       .uniq
       .sort
   end
+
+  def excluded_courses
+    # Does the Candidate have any submitted Applications?
+    return unless candidate.application_choices
+                           .joins(:application_form)
+                           .where(application_form: { recruitment_cycle_year: current_year })
+                           .exists?(status: ApplicationStateChange::STATES_VISIBLE_TO_PROVIDER)
+
+    # What Courses has the Candidate applied for?
+    # course codes & provider codes
+    candidate.application_choices
+                       .joins(:application_form)
+                       .where(application_form: { recruitment_cycle_year: current_year })
+                       .where(status: ApplicationStateChange::STATES_VISIBLE_TO_PROVIDER)
+                       .joins(course_option: { course: :provider })
+                       .pluck('course.code', 'provider.code')
+                       .compact_blank
+                       .uniq
+                       .sort
```

Comment:
```suggestion
    candidate.current_application.application_choices.visible_to_provider
                       .joins(course_option: { course: :provider })
                       .pluck('course.code', 'provider.code')
                       .compact_blank
                       .uniq
                       .sort
```

Is there a reason not to use existing scopes for grabbing these choices?

Reply:
Nice catch - looks like I was just trying to make my life difficult 😝